ebpf/aya-log-ebpf-macros/src/expand.rs

```diff
@@ -0,0 +1,192 @@
+                input.parse::<Token![:]>()?;
+                let t: Expr = input.parse()?;
+                input.parse::<Token![,]>()?;
+                Some(t)
+            }
+            false => None,
+        };
+
+        let lvl: Expr = input.parse()?;
+        input.parse::<Token![,]>()?;
+        let format_string: Expr = input.parse()?;
+
+        // Parse variadic arguments.
+        let formatting_args: Option<Punctuated<Expr, Token![,]>> = if input.is_empty() {
+            None
+        } else {
+            input.parse::<Token![,]>()?;
+            Some(Punctuated::parse_terminated(input)?)
+        };
+
+        Ok(Self {
+            ctx: ctx,
+            target: target,
+            lvl: lvl,
+            format_string: format_string,
+            formatting_args: formatting_args,
+        })
+    }
+}
+
+pub(crate) fn log(args: LogArgs) -> Result<TokenStream> {
+    let ctx = args.ctx;
+    let target = match args.target {
+        Some(t) => quote! { #t },
+        None => quote! { module_path!() },
+    };
+    let lvl = args.lvl;
+    let format_string = args.format_string;
+
+    let (num_args, quote_args) = match args.formatting_args {
+        Some(formatting_args) => {
+            let formatting_exprs = formatting_args.iter();
+            let num_args = formatting_exprs.len();
+
+            let quote_args = quote! {{
+                use ::aya_log_ebpf::WriteToBuf;
+                Ok::<_, ()>(record_len) #( .and_then(|record_len| {
+                    { #formatting_exprs }.write(&mut buf.buf[record_len..]).map(|len| record_len + len)
+                }) )*
+            }};
+
+            (num_args, quote_args)
+        }
+        None => (0, quote! {}),
+    };
+
+    Ok(quote! {
+        {
+            if let Some(buf) = unsafe { ::aya_log_ebpf::AYA_LOG_BUF.get_mut(0) } {
+                if let Ok(header_len) = ::aya_log_ebpf::write_record_header(
+                    &mut buf.buf,
+                    #target,
+                    #lvl,
+                    module_path!(),
+                    file!(),
+                    line!(),
+                    #num_args,
+                ) {
+                    if let Ok(message_len) = ::aya_log_ebpf::write_record_message(
+                        &mut buf.buf[header_len..],
+                        #format_string
+                    ) {
+                        let record_len = header_len + message_len;
+
+                        if let Ok(record_len) = #quote_args {
+                            if record_len <= ::aya_log_ebpf::macro_support::LOG_BUF_CAPACITY {
+                                let _ = unsafe { ::aya_log_ebpf::AYA_LOGS.output(
+                                    #ctx,
+                                    &buf.buf[..record_len], 0
+                                )};
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    })
+}
+
+pub(crate) struct LogLvlArgs {
```

Comment:
It seems a bit ugly to have these almost identical LogArgs and LogLvlArgs. I think they should either be merged somehow, or we should only have `LogLevelArgs` (two extra characters for readability :P), have `log(args: LogLevelArgs)` then parse the level, and pass it to `log_impl(level, args)`. `warn()` `error()` etc then would similarly do `log_impl(Level::Error, args)`, `log_impl(Level::Warn, args)` etc.

Reply:
Done, I managed to merge everything into `LevelArgs` and handle `level` as an optional argument.